python-frontend/src/main/java/org/sonar/python/caching/Caching.java

```
@@ -0,0 +1,109 @@
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.sonar.plugins.python.api.caching.CacheContext;
+import org.sonar.python.index.AmbiguousDescriptor;
+import org.sonar.python.index.ClassDescriptor;
+import org.sonar.python.index.Descriptor;
+import org.sonar.python.index.DescriptorUtils;
+import org.sonar.python.index.FunctionDescriptor;
+import org.sonar.python.index.VariableDescriptor;
+import org.sonar.python.types.protobuf.DescriptorsProtos;
+
+public class Caching {
+
+  private final CacheContext cacheContext;
+
+  public static final String INPORT_MAP_CACHE_KEY_PREFIX = "python_import_map:";
+  public static final String PROJECT_SYMBOL_TABLE_CACHE_KEY_PREFIX = "python_project_symbol_table:";
+
+  public Caching(CacheContext cacheContext) {
+    this.cacheContext = cacheContext;
+  }
+
+  public void writeImportMapEntry(String moduleFqn, Set<String> imports) {
+    byte[] importData = String.join(";", imports).getBytes(StandardCharsets.UTF_8);
+    String cacheKey = INPORT_MAP_CACHE_KEY_PREFIX + moduleFqn;
+    cacheContext.getWriteCache().write(cacheKey, importData);
+  }
+
+  public void writeProjectLevelSymbolTableEntry(String moduleFqn, Set<Descriptor> descriptors) {
+    String cacheKey = PROJECT_SYMBOL_TABLE_CACHE_KEY_PREFIX + moduleFqn;
+    cacheContext.getWriteCache().write(cacheKey, moduleDescriptor(descriptors).toByteArray());
+   }
+
+  public Optional<Set<Descriptor>> readProjectLevelSymbolTableEntry(String moduleFqn) {
+    cacheContext.getReadCache().contains(moduleFqn);
+    byte[] bytes = cacheContext.getReadCache().readBytes(moduleFqn);
+    if (bytes == null) {
+      return Optional.empty();
+    }
+    try {
+      return Optional.of(DescriptorUtils.deserializeProtobufDescriptors(bytes));
+    } catch (InvalidProtocolBufferException e) {
+      e.printStackTrace();
```

I think we may want to re-throw an exception here.

---

Actually I think we can log something here and return null. The caller will know the cache entry is missing/corrupted through the null return.